helix-dap/src/client.rs

```diff
@@ -0,0 +1,633 @@
+struct VariablesResponseBody {
+    variables: Vec<Variable>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OutputEventBody {
+    pub output: String,
+    pub category: Option<String>,
+    pub group: Option<String>,
+    pub line: Option<usize>,
+    pub column: Option<usize>,
+    pub variables_reference: Option<usize>,
+    pub source: Option<Source>,
+    pub data: Option<Value>,
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StoppedEventBody {
+    pub reason: String,
+    pub description: Option<String>,
+    pub thread_id: Option<usize>,
+    pub preserve_focus_hint: Option<bool>,
+    pub text: Option<String>,
+    pub all_threads_stopped: Option<bool>,
+    pub hit_breakpoint_ids: Option<Vec<usize>>,
+}
+
+#[derive(Debug)]
+pub struct Client {
+    id: usize,
+    _process: Option<Child>,
+    server_tx: UnboundedSender<Request>,
+    request_counter: AtomicU64,
+    capabilities: Option<DebuggerCapabilities>,
+    awaited_events: Arc<Mutex<HashMap<String, Sender<Event>>>>,
+}
+
+impl Client {
+    pub fn streams(
+        rx: Box<dyn AsyncBufRead + Unpin + Send>,
+        tx: Box<dyn AsyncWrite + Unpin + Send>,
+        id: usize,
+        process: Option<Child>,
+    ) -> Result<Self> {
+        let (server_rx, server_tx) = Transport::start(rx, tx, id);
+
+        let client = Self {
+            id,
+            _process: process,
+            server_tx,
+            request_counter: AtomicU64::new(0),
+            capabilities: None,
+            awaited_events: Arc::new(Mutex::new(HashMap::default())),
+        };
+
+        tokio::spawn(Self::recv(Arc::clone(&client.awaited_events), server_rx));
+
+        Ok(client)
+    }
+
+    async fn recv(
+        awaited_events: Arc<Mutex<HashMap<String, Sender<Event>>>>,
+        mut server_rx: UnboundedReceiver<Payload>,
+    ) {
+        while let Some(msg) = server_rx.recv().await {
+            match msg {
+                Payload::Event(ev) => {
+                    let name = ev.event.clone();
+                    let hashmap = awaited_events.lock().await;
+                    let tx = hashmap.get(&name);
+
+                    match tx {
+                        Some(tx) => match tx.send(ev).await {
+                            Ok(_) => (),
+                            Err(_) => error!(
+                                "Tried sending event into a closed channel (name={:?})",
+                                name
+                            ),
+                        },
+                        None => {
+                            info!("unhandled event");
+                            // client_tx.send(Payload::Event(ev)).expect("Failed to send");
+                        }
+                    }
+                }
+                Payload::Response(_) => unreachable!(),
+                Payload::Request(_) => todo!(),
```

Comment:
Even though we specify `"supportsRunInTerminalRequest":false` when connecting, codelldb will give this request to us, causing crash here.
```json
{"type":"request","seq":2,"command":"runInTerminal","arguments":{"args":["/.../adapter/codelldb","terminal-agent","--port=36691"],"cwd":"","kind":"integrated","title":"Debug"}}
```

Reply:
I'll probably try implementing that using the user. I hope LLDB will wait for user to start stuff in their other terminal. Looks like it won't work with a dummy impl.

Reply:
Actually I got this later, so we can probably set some config for that:
```
> [stderr] Failed to redirect stdio to a terminal. (Terminal agent did not respond within the allotted time.)
```

Reply:
Note: `console: "internalConsole"` argument to `launch` will redirect output to DAP `output` messages

Reply:
Can you test against the adapter provided by LLDB upstream (lldb-vscode) instad of vscode-lldb? It's the one that emacs dap-mode also supports: https://github.com/llvm/llvm-project/tree/main/lldb/tools/lldb-vscode

Reply:
I'll do that later, since I figured out how to deal with codelldb quirks

Reply:
vscode-lldb is a 3rd party plugin and specifically targets VSCode. Since it's quirky and doesn't follow the spec, I think it's fine to completely ignore and only support the official adapter.

Reply:
ok

Reply:
For reference: emacs dap-mode seems to only support lldb-vscode, whereas nvim-dap supports lldb-vscode and vscode-cpptools: https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#ccrust-via-vscode-cpptools

Maybe their configs can help you out a bit

Reply:
Thx, extra config wasn't necessary. See examples, they're pretty much identical except for running go/c debugger, debuggee and setting breakpoints on different lines.